source/applications/psyllid.cc

```diff
@@ -35,33 +38,26 @@ int main( int argc, char** argv )
     {
         // The application
         scarab::main_app the_main;
-        run_server the_server;
+        conductor the_conductor;
+        the_conductor.set_rc_creator< daq_control >();
```

Comment:
This comment is actually not about psyllid, but about sandfly.

Anyway, I don't get why the conductor class has this additional detour for setting the function `f_rc_creator` . Why not just a template class in the first place? Are there plans to change the rc_type dynamically? I didn't find an example of that. Or is the idea to prevent instantiation of the template with anything that is not derived from `run_control` ? If so, there is probably a solution using type traits.

Reply:
I really should have left more comments to explain what I was thinking here, but I think my reasoning was:

* I didn't want to make the whole class a template class because I didn't want to put all of the class in the header file.  Since conductor.cc "knows" about almost all of the control classes, I think I was concerned about the increased risk of a circular-include situation.  I don't know whether this was just a worry or an actual problem that I was trying to avoid.  At least since `conductor.hh` is only included from executables there's a little less risk, and it might actually be possible.
* The `run_control` object needed to be created when `conductor::execute()` was called, not when `conductor` was executed in the application callback function.  I considered making `conductor::execute()` a template function instead, but that has the same issue with include files as the point above.

There's no intention to make `f_rc_creator` change dynamically.

Reply:
There are ways around that. You can do an explicit template specialization in the cpp file to still get a separation in h and cpp file. But that requires of course that all possible instances are known in advance. I get that with outsourcing the conductor to its own library not all possible instances will be known by design. I also assume that making the class abstract and inheriting from that are excluded due to the runtime overhead of that.

In this case I think you could solve it like here https://stackoverflow.com/a/41292751 .
The first two files mentioned in that post are then part of sandfly, while the third one is part of psyllid and does the explicit template specialization for `daq_control`. It might be an unusual solution, but personally I think it makes the conductor class much clearer. The way it is `f_rc_creator` is fairly complicated.